File: jay6627_lab01.py
# Name: Jack Yohn
# 1000#: 1002056627
# Lang/Ver: Python 3.11
# OS: Linux (Omega server)
#
# Uses a recursive function with os.scandir.
# Skips symlinks to avoid infinite loops. Prints only the byte total.

import os
import sys

def sum_dir(path: str) -> int:
    """Recursively sum regular file sizes under 'path'."""
    total = 0
    try:
        # If it's a symlink, skip it entirely
        if os.path.islink(path):
            return 0
        # If it's a file, return its size via lstat (do not follow links)
        st = os.lstat(path)
        if os.path.isfile(path):
            return st.st_size
        # If it's a directory, recurse into entries
        if os.path.isdir(path):
            with os.scandir(path) as it:
                for entry in it:
                    # Skip "." and ".." implicitly; scandir won’t return them explicitly
                    child_path = entry.path
                    try:
                        # Use lstat to avoid following symlinks
                        if entry.is_symlink():
                            continue
                        if entry.is_file(follow_symlinks=False):
                            total += os.lstat(child_path).st_size
                        elif entry.is_dir(follow_symlinks=False):
                            total += sum_dir(child_path)  # recursion
                        # Other types ignored
                    except Exception:
                        # Skip unreadable entries
                        pass
        # If neither file nor directory, ignore
    except Exception:
        # Skip paths we can't stat/read
        return 0
    return total

if __name__ == "__main__":
    # Always start at current directory per assignment
    print(sum_dir("."))

File: jay6627_lab01.cpp
/* Name: Jack Yohn
 * 1000#: 1002056627
 * Lang/Ver: C++17
 * OS: Linux (Omega server)
 *
 * Recursively sum sizes of regular files starting at ".".
 * Skips symlinks to avoid loops and double-counting.
 * Prints only the integer byte total (no extra text).
 */

#define _XOPEN_SOURCE 700
#include <iostream>
#include <string>
#include <cstring>
#include <dirent.h>
#include <sys/stat.h>
#include <unistd.h>
#include <cerrno>
#include <limits.h>
#include <stdint.h>

static bool isDotOrDotDot(const char* name) {
    return std::strcmp(name, ".") == 0 || std::strcmp(name, "..") == 0;
}

static bool joinPath(const std::string& base, const char* name, std::string& out) {
    // Safe path join without exceeding PATH_MAX when converted to C string
    out.clear();
    if (!base.empty() && base.back() == '/')
        out = base + name;
    else if (base.empty())
        out = name;
    else
        out = base + "/" + name;

    // Rough guard: ensure it won’t overflow typical C buffers, though we use std::string
    return out.size() < PATH_MAX;
}

// Recursive function: Returns total size (bytes) of regular files under 'path'.

static uint64_t sumDir(const std::string& path) {
    uint64_t total = 0;

    DIR* dir = opendir(path.c_str());
    if (!dir) {
        // If not a directory or unreadable, try to count it as a single file.
        struct stat st;
        if (lstat(path.c_str(), &st) == 0) {
            if (S_ISREG(st.st_mode)) {
                total += static_cast<uint64_t>(st.st_size);
            }
        }
        return total;
    }

    // Ensure directory stream gets closed even if we early-return
    struct DirCloser {
        DIR* d;
        ~DirCloser() { if (d) closedir(d); }
    } closer{dir};

    struct dirent* ent;
    std::string child;

    while ((ent = readdir(dir)) != nullptr) {
        if (isDotOrDotDot(ent->d_name)) continue;

        if (!joinPath(path, ent->d_name, child)) {
            // Path too long; skip safely
            continue;
        }

        struct stat st;
        if (lstat(child.c_str(), &st) != 0) {
            // Unreadable entry, skip
            continue;
        }

        if (S_ISLNK(st.st_mode)) {
            // Skip symlinks entirely
            continue;
        } else if (S_ISREG(st.st_mode)) {
            total += static_cast<uint64_t>(st.st_size);
        } else if (S_ISDIR(st.st_mode)) {
            total += sumDir(child); // recursion
        }
    }

    return total;
}

int main() {
    uint64_t total = sumDir(".");
    std::cout << total << '\n'; // print ONLY the integer
    return 0;
}
